### PR TITLE
[no_early_kickoff] [data] Improve our handling of tensor returns in strict mode

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

In previous PRs, we automatically converted lists into numpy arrays. This PR generalizes this to other array-like types, namely:
- Array-like objects (those implementing `__array__`) are also treated as tensors
- Lists of Array-like objects are also treated as tensors
- Ragged arrays of these types are automatically created, without needing to call `create_ragged_ndarray` manually
- Support this for `map` in addition to `map_batches`

Before (Torch map_batches UDF example):
```
    batch = collections.defaultdict(list)
    for output in outputs:
        for key, value in output.items():
            batch[key].append(value.cpu().detach().numpy())
    for key, value in batch.items():
        batch[key] = create_ragged_ndarray(value)
```

After:
```
    batch = collections.defaultdict(list)
    for output in outputs:
        for key, value in output.items():
            batch[key].append(value.cpu().detach())
```
Note that Torch still requires users to call detach/cpu, per: https://github.com/pytorch/pytorch/issues/36560

This is a breaking API change, so it is scoped to strict mode.

### CODE REVIEW

**Review by raulchen — APPROVED** [review]

File: python/ray/data/_internal/numpy_support.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            # Try to cast the inner scalars to numpy as well, to avoid unnecessarily
+            # creating an inefficient array of array of object dtype.
+            if all(is_valid_udf_return(e) for e in udf_return_col):
+                udf_return_col = [np.array(e) for e in udf_return_col]
+            shapes = set()
+            if all(isinstance(e, np.ndarray) for e in udf_return_col):
+                for e in udf_return_col:
+                    shapes.add(e.shape)
+            if len(shapes) > 1:
+                logger.get_logger().info(
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/_internal/numpy_support.py:60]: I think this is an important warning message, and it will be deduplicated across workers. I'm not worried about perf as we do multiple sequential scans across the batch already, and any users worried about perf can return a contiguous array.

File: python/ray/data/_internal/numpy_support.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        return udf_return_col
+
+    if isinstance(udf_return_col, list):
+        # Try to convert list values into an numpy array via
+        # np.array(), so users don't need to manually cast.
+        # NOTE: we don't cast generic iterables, since types like
+        # `str` are also Iterable.
+        try:
+            # Try to cast the inner scalars to numpy as well, to avoid unnecessarily
+            # creating an inefficient array of array of object dtype.
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/_internal/numpy_support.py:52]: It happens for [torch.Tensor(), torch.Tensor(), ...].

**Comment by ollie-iterators** [comment]
If this is only happening in strict mode, then there should be a note about removing the condition for strict mode when strict mode is unable to be disabled.

**Comment by ericl** [comment]
:+1: Added a note to the docs for this.

File: python/ray/data/_internal/numpy_support.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (raulchen) [reply python/ray/data/_internal/numpy_support.py:63]: how does this get duplicated? it seems to me that this will print for every batch.

[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/_internal/numpy_support.py:63]: Oh good point, maybe this is too much even with the log monitor deduplication. I'll remove it.

**Review by pcmoritz — APPROVED** [review]
I think this is a good idea -- let's also make sure to keep an eye out if there are cases where it might break or cause undesired behavior :)

File: doc/source/data/working-with-tensors.rst [hunk 3]
```diff
@@ -69,6 +69,42 @@ Elements of ragged arrays are regular ndarrays.
...
+
+    def batch_increase_brightness(batch: Dict[str, np.ndarray]) -> Dict:
+        batch["image"] = np.clip(batch["image"] + 4, 0, 255)
+        return batch
+
+    # Increase the brightness, batch at a time.
+    ds.map_batches(batch_increase_brightness)
+
+In this example, we return ``np.ndarray`` directly as the output. Ray Data will also treat
+returned lists of ``np.ndarray`` and objects implementing ``__array__`` as tensor data.
```
Comment (amogkam) [inline doc/source/data/working-with-tensors.rst:103]: add some examples of object implementing the array protocol (pytorch tensors, etc.)?

I don't think many people know what this actually means

File: python/ray/data/_internal/numpy_support.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            if all(is_valid_udf_return(e) for e in udf_return_col):
+                udf_return_col = [np.array(e) for e in udf_return_col]
+            shapes = set()
+            if all(isinstance(e, np.ndarray) for e in udf_return_col):
+                for e in udf_return_col:
+                    shapes.add(e.shape)
+            if len(shapes) > 1:
+                # This util works around some limitations of np.array(dtype=object).
+                udf_return_col = create_ragged_ndarray(udf_return_col)
```
Comment (amogkam) [inline python/ray/data/_internal/numpy_support.py:64]: nice, very clever!

File: python/ray/data/_internal/numpy_support.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            else:
+                udf_return_col = np.array(udf_return_col)
+        except Exception as e:
+            raise ValueError(
+                "Failed to convert column values to numpy array: "
+                f"({_truncated_repr(udf_return_col)}): {e}."
+            )
+    elif hasattr(udf_return_col, "__array__"):
+        # Converts other array-like objects such as torch.Tensor.
```
Comment (amogkam) [inline python/ray/data/_internal/numpy_support.py:73]: we don't need to check the inner elements for this case?

File: python/ray/data/tests/test_numpy_support.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+def test_list_of_numpy_scalars(ray_start_regular_shared):
+    data = [np.int64(1), np.int64(2), np.int64(3)]
+    output = do_map_batches(data)
+    assert_structure_equals(output, np.array([1, 2, 3], dtype=np.int64))
+
+
+def test_list_of_objects(ray_start_regular_shared):
+    data = [1, 2, 3, UserObj()]
+    output = do_map_batches(data)
+    assert_structure_equals(output, np.array([1, 2, 3, UserObj()]))
```
Comment (amogkam) [inline python/ray/data/tests/test_numpy_support.py:45]: check that this is object dtype right?

**Review by amogkam — COMMENTED** [review]
this is awesome!

**Review by amogkam — COMMENTED** [review]
this is awesome!

File: python/ray/data/tests/test_numpy_support.py (hunk 6 above)
Comment (ericl) [reply python/ray/data/tests/test_numpy_support.py:45]: Yup, assert struct equals will check the types are exactly equal. The literal on the right returns an array with object dtype.

File: python/ray/data/_internal/numpy_support.py (hunk 5 above)
Comment (ericl) [reply python/ray/data/_internal/numpy_support.py:76]: In this case, the whole thing is already an array since it implements the array protocol, and hence is guaranteed to be convertible to a numpy array without fuss.
